magenta-lib/src/main/scala/magenta/graph/Graph.scala

```diff
@@ -74,29 +80,49 @@ case class Graph[T](edges: Set[Edge[T]]) {
   }
 
   def joinSeries(other: Graph[T]): Graph[T] = {
-    val ourEndEdges = edges.filter(_.to == EndNode)
-    val otherStartEdges = other.edges.filter(_.from == StartNode)
-    val joiningEdges = ourEndEdges.flatMap { endEdge =>
-      otherStartEdges.map(startEdge => Edge(endEdge.from, startEdge.to, startEdge.priority))
+    val ourEndEdges = incoming(EndNode)
+    val otherStartEdges = other.orderedOutgoing(StartNode)
+    val joiningEdges = ourEndEdges.foldLeft(edges) { case (acc, endEdge) =>
+      val sourceNodeOutgoing = orderedOutgoing(endEdge.from)
+      // build new outgoing nodes
+      val before = sourceNodeOutgoing.takeWhile(endEdge != _)
+      val newEdges = otherStartEdges.map(startEdge => Edge(endEdge.from, startEdge.to, before.size + startEdge.priority))
+      val after = sourceNodeOutgoing.dropWhile(endEdge != _).tail
+      val afterReprioritised = after.map(_.incPriority(newEdges.size - 1))
+      acc - endEdge -- after ++ newEdges ++ afterReprioritised
     }
-    val mergedEdges = (edges -- ourEndEdges) ++ (other.edges -- otherStartEdges) ++ joiningEdges
+    val mergedEdges = joiningEdges ++ (other.edges -- otherStartEdges)
     Graph(mergedEdges)
   }
 
-  def toList: List[T] = {
-    def traverseFrom(node: Node[T], visited: Set[Node[T]]): List[Node[T]] = {
-      val predecessors: Set[Node[T]] = this.predecessors(node)
-      if ((predecessors -- visited).nonEmpty) {
-        // if there are some predecessors of this node that we haven't yet visited then return empty list - we'll be back
-        Nil
+  lazy val nodeList: List[Node[T]] = {
```

Comment:
This is now a more typical topological sort of the graph - I switched this from recording visited nodes to recording traversed edges which is more important in many shapes of graphs.